### ISSUE TITLE
Slowdown

### ISSUE THREAD

**Issue by aPaci95** [eshapard/AnkiHabitica#34]
Hello, every time I enter in a deck (after i installed this addon) it takes 20-30 seconds to load the questions; before it used to take 1 second, maybe 2 for bigger decks. What can I do?

**Comment by twilsonco** [comment]
That's weird. This add-on only does stuff when you answer a card or sync. Otherwise none of its functions are accessed, so that makes no sense.
So you can remove the add-on and the speed returns to normal, and re-add the add-on and the delay happens? Every time?

**Comment by aPaci95** [comment]
Yes. Today for example I retried to install the addon. When I opened anki after the restart it freezed (I couldn't do anything for something like 20-30 seconds). Then it asked me if i wanted to sync my past reviews with habitica, i clicked yes and then it freezed for another 15-20 seconds. Then appeared another window where it said he was "scoring points to habitica". This taked a 1-2 minutes. After it reached 100% it freezed for another 1-2 minutes. Now it works fine, but everytime I enter in a deck and I click on study now it takes 20-30 seconds to load the first question, and when i answer it takes another 20-30 seconds to load the second question, and so on.

**Comment by twilsonco** [comment]
What kind of machine are you running on? This sounds like really poor performance for operations that take ~0.5s on my machine which isn’t even that fast (dual core 2.5GHz and 8GB memory).

**Comment by eshapard** [comment]
Opening a deck causes the progress bar to be built which requires
contacting the habitica server. That takes some time. This is almost
certainly the issue. After opening the deck, most of the calls to habitica
are in the background, but this first one isn't.

It's actually the loading of that first card that triggers it; not opening
a deck.

You could try turning off show_mini_stats. That might speed things up a
little.

**Comment by aPaci95** [comment]
Intel core i5-4300U 1.90 GHz
4 GB RAM
SSD

What can I do to improve my performances? Right now it is almost impossible to use this addon (for me)

**Comment by twilsonco** [comment]
In your anki add ons folder you’ll find a folder for AnkiHabitica. In there you’ll find log files like AnkiHabitica.log[.x]. Close Anki, erase all the log files, then open it back up and go through a few cards with this add-on, then post the log file here.

**Comment by eshapard** [comment]
You may have to set keep_log = True in the main ankihabitica.py file
first... the logging function was throwing some errors every once in a
while, so I set it to be turned off by default.

**Comment by eshapard** [comment]
How does this make it almost impossible to use the addon? From what you
described, you only see a slowdown when you first open a deck. Do you have
a ton of decks or something?

**Comment by eshapard** [comment]
Oh, I see, you have a big lag before each card. Not sure what could cause
this. How's your connection speed? Maybe the logs will give us some clues...

**Comment by twilsonco** [comment]
Since everything is timestamped, I expect we’ll be able to see what the problem is pretty easily.

**Comment by eshapard** [comment]
One more thought... it's happened a couple times that someone had another
addon that conflicted with ankihabitica. You might try un-installing all
other addons just to see if that makes a difference. If that seems to fix
the issue, you can re-install your other addons one by one.

**Comment by twilsonco** [comment]
This is much less of a pain than it sounds. Just move the contents of the Anki addons folder somewhere else with the exception of the AnkiHabitica.py and AnkiHabitica folder. Then restart Anki. When you’re done testing, move everything back one add-on at a time.

**Comment by steelwolf** [comment]
I'm having this same issue, with each new card taking approximately 20 seconds to load. I got rid of all my other addons using the method described above but that has not changed anything.

**Comment by twilsonco** [comment]
Please post some logs (see above) so that we can get a better idea of what’s happening.

**Comment by eshapard** [comment]
OK, well the next step would be to edit the addon to turn on logging and
then post the log somewhere. The log will be created in the AnkiHabitica
directory after you turn on logging.

**Comment by steelwolf** [comment]
Sorry for the delay; it took a little while to do some cards and then remove info from the logs. The paste will expire in a day in case I missed something.

[http://pastebin.com/8KCUCjbb](url)

**Comment by twilsonco** [comment]
OK, so it’s the counting of finished decks that’s doing it:

2017-02-01 10:53:13,080 [    MainThread:      db_helper.py:  119 -                    decks_count()]    DEBUG: Begin function
2017-02-01 10:53:41,664 [    MainThread:      db_helper.py:  197 -                    decks_count()]    DEBUG: End function returning: 1

28 seconds is way too long for that step.

**Comment by steelwolf** [comment]
FWIW I only did about 5 cards for the purposes of this test. I also have a ton of individual decks in my total Anki collection; however, I studied from a single deck the whole time. It does make sense to have a bit of a load time when first opening Anki and/or when opening a deck, but the 20-30 second load time between every card is odd to me.

**Comment by twilsonco** [comment]
Could we call decks_count() in a new thread and just have it update the score once it finishes? It would result in you getting the points for completing a deck later, but it could help performance.

**Comment by eshapard** [comment]
Just noticed that too.

Do you have a lot of decks?

What we can do is disable that check. You won't get points for finishing
all the reviews in a deck, but you won't be waiting 28 seconds.

I don't have time to look into making changes right now, but I can alter
the code so you can disable this check later.

If you know a little coding, you can do this yourself. Look for a function
that calculates your total score by adding up different points and disable
the decks_count() part.

**Comment by eshapard** [comment]
Michael: The addon can't tell which deck you've been working in, so it has
to check all of them. Since you have a lot of decks, that's what's causing
the slowdown.

Tim: Possibly... That's a good idea. Do you think it might lock up the
database though? Disabling that check seems like the most error-prone way
to do this.

**Comment by twilsonco** [comment]
All the DB stuff is read only. Should be able to do that anytime as there’s no thread locking with SQLite. Worst case scenario is that a DB query misses a change that was being written at the same time, but for this add-on that’s hardly a big deal, as the change would be caught on the next query.

At 28s per call, there’s an issue with the threads running decks_count() piling up…

A more thorough solution would be to only check decks that are associated with the card that was just answered, so its deck and all parent decks. This would be the function to use for hrpg_realtime, and the full decks_count() would then only be used for checking after a sync.

**Comment by steelwolf** [comment]
That was it. I commented out `dbdecks` from the `calculate_db_score` function and everything is pretty snappy. 

Subdecks are a decent way of organizing different categories of cards but it definitely bloats up quickly.

**Comment by eshapard** [comment]
OK, I think I'll code it so that you can disable any of the various checks
that go into the total score... that will be an easy change and will fix
this issue as long as a user disables the decks_count check.

Maybe look into one of the other options later, but I'm swamped with work
these days.

**Comment by eshapard** [comment]
Good work!

Yeah, Damien says not to have many decks, but most people don't experience
a slowdown until they get several dozens. This particular function just
happens to be very sensitive to deck number.

It's a very difficult proposition to try to figure out how many decks were
completed each day, as Anki doesn't store these statistics; the function
has to do a lot of detective work.

**Comment by twilsonco** [comment]
If you recall the way I had tweaked the setup in my old version of the add-on. Present the user with a series of dialogs to set the number of events required to up their score (e.g. 5 correct answers, 10 incorrect answers), and give them the option of putting in 0 to disable that type of scoring completely. That way a user can enable/disable an event from Anki.

**Comment by twilsonco** [comment]
A lot of very sexy detective work, if I may say  so myself!

**Comment by twilsonco** [comment]
Also, I still like the idea of using Anki events wherever possible, and then reconciling the score from events with the score in the database infrequently. Would basically eliminate the runtime of HRPG.

**Comment by eshapard** [comment]
Yeah, I think we thought of that when we were developing this... Good idea,
but I focused on just getting the database stuff to work at the time...
simpler code that way.

I guess now that the code is working well, we could look at event-driven
actions again. Maybe only run the database stuff once (on start, and/or via
the menu) and compare it to the score log; highest score wins. The rest of
the time, go purely off of events.

The events should be pretty easy to code with hooks and all (except for the
darn finished deck one). I don't really have the energy for it right
now.... you feel like doing it? Maybe a progress bar to show that we're
working on calculating the score from the database.

A gui to choose config options is a good idea. I envisioned having a custom
window pop up with all the settings there, but didn't want it badly enough
to put in the work to figure out how to do that.

And yes, that is some sexy detective work ;-)

**Comment by twilsonco** [comment]
I think the old version of the add-on has most of the code we need. Database stuff would only be done on sync. 

As I think about it more, I’m not sure finishing a deck actually deserves points, since you’re getting points for each answer anyways. Doesn’t make sense that all decks should be equal regardless of size, and if you had to do more cards and earn more points to finish a large deck anyways, why give extra points for it? Would make more sense to have either card-based or deck-based scoring, but not both.

With that in mind, we should probably just deprecate deck-based scoring, unfortunately losing all that sexy code…. Or just disable it by default since it’s proving to be a performance hit.

**Comment by eshapard** [comment]
Yeah, I originally intended it to be an incentive to finish all the reviews
in a deck, like a reward for reaching the finish line. But treating all
decks equally can lead to some strange results...

I vote for disabling by default and let more advanced users edit the code
to enable it themselves.

**Comment by eshapard** [comment]
I updated the github code so that by default, deckpoints is set to zero, and any points setting set to zero disables the related database check.... Simple change, but I haven't tested it out yet.

Will update ankiweb version after I'm sure I didn't screw anything up.

**Comment by twilsonco** [comment]
Regarding providing a single gui to get all the necessary options for AnkiHabitica, check out easygui.multenterbox() http://easygui.sourceforge.net/tutorial.html#multenterbox <http://easygui.sourceforge.net/tutorial.html#multenterbox>
easygui has no dependencies and comes with most python distros, so at most you can just throw a few python files into a subfolder like we do for logging, and then the gui is no problem.

**Comment by eshapard** [comment]
OK, we can play with that if either one of us finds the time... my experience with guis is that they aren't all that hard, but I rarely code anything that uses them, so I have to start from scratch each time.

**Comment by eshapard** [comment]
Updated the addon in AnkiWeb today.
